c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -388,25 +388,36 @@ class THCCachingAllocator {
     return basePtr;
   }
 
-  void recordStream(void* ptr, cuda::CUDAStream stream) {
+  void recordStream(const DataPtr& ptr, cuda::CUDAStream stream) {
     // Empty tensor's storage().data() might be a null ptr. As there is no
     // blocks associated with those tensors, it is fine to do nothing here.
-    if (!ptr) {
+    if (!ptr.get()) {
       return;
     }
 
+    // The ptr is not allocated by CUDA, most likely user error
+    TORCH_CHECK(
+        ptr.device().type() == kCUDA, "Tensor is not allocated in CUDA");
+
+    // If a tensor is not allocated by this instance, simply skip
+    // This usually happens when CUDA tensors are shared across processes,
+    // we have implemented reference counting based sharing mechanism to
+    // guarantee tensors won't be accidentally freed by one process while
+    // they are still being used in another
+    if (ptr.get_deleter() != &raw_delete)
+      return;
+
     std::lock_guard<std::recursive_mutex> lock(mutex);
-    Block* block = find_allocated_block(ptr);
-    // block could be nullptr in some cases, e.g., tensor loaded from blob, or
-    // shared from another process, or not pointing to a CUDA tensor.
-    if (block) {
-      if (stream.stream() == block->stream) {
-        // ignore uses on the allocation stream, since those don't require any
-        // special synchronization
-        return;
-      }
-      block->stream_uses.insert(stream);
+
+    Block* block = find_allocated_block(ptr.get());
+    // block must not be null reaching here
+    TORCH_INTERNAL_ASSERT(block != NULL, "No allocated block can be found");
```

Comment:
nit: nullptr rather than NULL in modern C++

Reply:
Okay, will update it.